src/functions/runtimeConfigExport.ts

```diff
@@ -0,0 +1,222 @@
+  /* prettier-ignore */
+  const baseKey = configKey
+      .toUpperCase()       // 1. Uppercase all characters (e.g. SOME-SERVICE.KEY)
+      .replace(/\./g, "_") // 2. Dots to underscores (e.g. SOME-SERVICE_KEY)
+      .replace(/-/g, "_"); // 3. Dashses to underscores (e.g. SOME_SERVICE_KEY)
+
+  let envKey = baseKey;
+  try {
+    env.validateKey(envKey);
+  } catch (err) {
+    if (err instanceof env.KeyValidationError) {
+      envKey = prefix + envKey;
+      env.validateKey(envKey);
+    }
+  }
+  return envKey;
+}
+
+/**
+ * Convert runtime config into a map of env vars.
+ */
+export function configToEnv(configs: Record<string, unknown>, prefix: string): ConfigToEnvResult {
+  const success = [];
+  const errors = [];
+
+  for (const [configKey, value] of Object.entries(flatten(configs))) {
+    try {
+      const envKey = convertKey(configKey, prefix);
+      success.push({ origKey: configKey, newKey: envKey, value: value as string });
+    } catch (err) {
+      if (err instanceof env.KeyValidationError) {
+        errors.push({
+          origKey: configKey,
+          newKey: err.key,
+          err: err.message,
+          value: value as string,
+        });
+      } else {
+        throw new FirebaseError("Unexpected error while converting config", {
+          exit: 2,
+          original: err,
+        });
+      }
+    }
+  }
+  return { success, errors };
+}
+
+/**
+ * Fill in environment variables for each project by converting project's runtime config.
+ *
+ * @return {ConfigToEnvResult} Collection of successful and errored conversion.
+ */
+export function hydrateEnvs(pInfos: ProjectConfigInfo[], prefix: string): string {
+  let errMsg = "";
+  for (const pInfo of pInfos) {
+    const { success, errors } = configToEnv(pInfo.config!, prefix);
+    if (errors.length > 0) {
+      const msg =
+        `${pInfo.projectId} ` +
+        `${pInfo.alias ? "(" + pInfo.alias + ")" : ""}:\n` +
+        errors.map((err) => `\t${err.origKey} => ${clc.bold(err.newKey)} (${err.err})`).join("\n") +
+        "\n";
+      errMsg += msg;
+    } else {
+      pInfo.envs = success;
+    }
+  }
+  return errMsg;
+}
+
+function escape(s: string): string {
+  // Escape newlines and tabs
+  const result = s
+    .replace("\n", "\\n")
+    .replace("\r", "\\r")
+    .replace("\t", "\\t")
+    .replace("\v", "\\v");
+  return result.replace(/(['"])/g, "\\$1");
+}
+
+/**
+ * Convert env var mapping to  dotenv compatible string.
+ */
+export function toDotenvFormat(envs: EnvMap[], header = ""): string {
+  const lines = envs.map(({ newKey, value }) => `${newKey}="${escape(value)}"`);
+  const maxLineLen = Math.max(...lines.map((l) => l.length));
+  return (
+    `${header}\n` +
+    lines.map((line, idx) => `${line.padEnd(maxLineLen)} # from ${envs[idx].origKey}`).join("\n")
+  );
+}
+
+/**
+ * Create a dotenv file for each project's environment variables.
+ */
+export function writeDotenvFiles(
```

Comment:
Do we really need to write to a file here or can we just create a map of filename to file contents? That would let us use `askWriteProjectFile` in the actual command.

Reply:
One goal was to have writing set of dotenv files atomic, but I do see value of keeping our code simple. Changing as suggested.